File: reverse_linked_list.c
// 206. Reverse Linked List
#include "leetcode.h"

/*
 * given the 'head' of a singly linked list, reverse the list, and return the
 * reversed list.
 */

struct ListNode {
  int val;
  struct ListNode *next;
};

struct ListNode *reverseList(struct ListNode *head) {
  if (!head || !head->next)
    return head;
  struct ListNode *new = (struct ListNode *)malloc(sizeof(struct ListNode));
  new->val = head->val;
  new->next = NULL;
  head = head->next;
  while (head) {
    struct ListNode *tmp = new;
    new = head;
    head = head->next;
    new->next = tmp;
  }
  return new;
}

File: reverse_linked_list.py
# 206. Reverse Linked List

"""
given the 'head' of a singly linked list, reverse the list, and return the
reversed list.
"""


# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        prev = None
        while head:
            curr = head
            head = head.next
            curr.next = prev
            prev = curr
        return prev


if __name__ == "__main__":
    obj = Solution()
    print(obj.reverseList(head=[1, 2, 3, 4, 5]))
    print(obj.reverseList(head=[1, 2]))
    print(obj.reverseList(head=[]))
